[inductor] use eager stride for custom op if no tags

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #148367

Fix https://github.com/pytorch/pytorch/issues/148356 

This is some sort of short term fix to recover the default behavior to apply layout constraint for custom ops when there are no tags.

A longer term attempt to make sure Inductor always gets correct eager strides is here: https://github.com/pytorch/pytorch/pull/148104

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

shunting314 committed:
[inductor] use eager stride for custom op if no tags

[ghstack-poisoned]

zou3519 commented on torch/_inductor/graph.py:
```diff
@@ -1138,14 +1138,22 @@ def call_function(self, target: Callable, args: Any, kwargs: dict[str, Any]) ->
+                    if torch._library.utils.is_builtin(target):
+                        decided_constraint = require_contiguous  # type: ignore[assignment]
+                    else:
+                        decided_constraint = constrain_to_fx_strides  # type: ignore[assignment]
```
shouldn't this just be decided_constraint = None to trigger the following logic? https://www.internalfb.com/code/fbsource/[c74f1bf3aedec6b6eaf53562636d0efd8ec461e4]/fbcode/caffe2/torch/_inductor/lowering.py?lines=182%2C192

shunting314 committed:
Update on "[inductor] use eager stride for custom op if no tags"


Fix https://github.com/pytorch/pytorch/issues/148356 

This is some sort of short term fix to recover the default behavior to apply layout constraint for custom ops when there are no tags.

A longer term attempt to make sure Inductor always gets correct eager strides is here: https://github.com/pytorch/pytorch/pull/148104

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

eellison reviewed (commented):
I would kind of rather we fix the relevant weight norm kernels then land this PR which is a non fully featured version of what richard was doing in the other pr.

Giving aten kernels this extra stride constraint is bad for things like channels optimization.

eellison commented:
@shunting314 I don't think this is a general problem for aten ops. the error is only for specific backward ops, which assume as an invariant that the forward would have already applied the layout constraints.

shunting314 commented:
>  I don't think this is a general problem for aten ops. the error is only for specific backward ops, which assume as an invariant that the forward would have already applied the layout constraints.

yea, that's what I mean by 'general'. There can be new ops like this be added in future, right?

eellison commented:
@shunting314 we are not adding new core aten ops at any fast rate, especially with backward and unchecked invariants around striding.

shunting314 committed:
Update on "[inductor] use eager stride for custom op if no tags"


Fix https://github.com/pytorch/pytorch/issues/148356 

This is some sort of short term fix to recover the default behavior to apply layout constraint for custom ops when there are no tags.

A longer term attempt to make sure Inductor always gets correct eager strides is here: https://github.com/pytorch/pytorch/pull/148104

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

shunting314 commented:
Chatted with @eellison offline. The default layout constraint for implicit fallback ATEN ops now are only applied to bwd graph since fwd ATen ops should work with any strides.

eellison commented on torch/_inductor/graph.py:
```diff
@@ -1138,14 +1138,27 @@ def call_function(self, target: Callable, args: Any, kwargs: dict[str, Any]) ->
+                    # For ATen ops, only apply the constraint for backward
+                    # ops since fwd ops should work for any strides.
+                    if torch._library.utils.is_builtin(target) and self.is_backward:
+                        decided_constraint = require_contiguous  # type: ignore[assignment]
```
Hmm, i guess in the future we could change this to matching the fx strides.

zou3519 replied:
do you mean "needs_exact_strides"?

shunting314 committed:
Update on "[inductor] use eager stride for custom op if no tags"


Fix https://github.com/pytorch/pytorch/issues/148356 

This is some sort of short term fix to recover the default behavior to apply layout constraint for custom ops when there are no tags.

A longer term attempt to make sure Inductor always gets correct eager strides is here: https://github.com/pytorch/pytorch/pull/148104

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

shunting314 commented:
@pytorchbot merge

shunting314 commented:
@pytorchbot merge -i
